### DISCUSSION TITLE
Best way to move data between two databases using SQLAlchemy

### DISCUSSION THREAD

**Discussion by ironsoul** [sqlalchemy/sqlalchemy#7571] [Usage Questions] [answered]
I am quite new to SQLAlchemy and Python as a whole and I am looking for some advise. I am looking at moving data from one Postgres DB to a different Postgres DB. I will eventually be doing some ETL work but the main purpose right now is to move the data (lift and shift). Currently I am not sure how to optimize the insert. I have a Stack Overflow question [Stack Overflow](https://stackoverflow.com/questions/70652544/best-way-to-move-data-between-two-databases-using-sqlalchemy) 

The code that I am currently using is attached at the bottom.

I'm sure I have missed something in the docs, but I have been going through it for a while and haven't found an answer yet.

I have tested the same code on both 1.3 and 1.4 using the Core not the ORM.
1. Firstly I have mostly been using 1.3 should I move to 1.4?
2. Is the Core the correct path for what I want to do, or should I be using the ORM for this? (I have started today looking into the ORM but still very new)
3. Am I using the correct code to do what I want or is the issue there? (I heard that there might be flags I need to set to optimize the insert)
4. How do I optimize the INSERT using the Core?

I unfortunately can't use pyscopg2 Copy function because I do not have SuperUser access to the DB.

Currently I am testing on a small table of about 140k records and it takes about 3 minutes to move. This is quite slow. I am looking at eventually moving 20M± records a day.

I am using AWS sagemaker to run the job, its run on a ml.t2.medium (4GB memory), which I know has limited memory. I did try on a larger instance ml.t2.xlarge (16GB memory) with a batch size of 10k records at a time which took 180 seconds to load the 140k records. 

Any advise will be greatly appreciated. Thanks.

Code sample:

```py
conn_p = create_engine(--db connection string--, echo=True)
conn_sl = create_engine(--db connection string--, echo=False)


        #table_1 is the source table and table_2 is the destination. This is how I have defined them.
table_1 = table("table_name", column('*'))

        #table_2 needs a schema where table_1 does not.
meta = MetaData(schema = "schema_name")

        #table_2 has 52 Columns defined like this.
table_2 = Table(
                "table_name",
		meta,
		Column('column_name_1'),
		Column('column_name_2'),
		...etc
		)
				

truncate_query = sqlalchemy.text("TRUNCATE TABLE schema.table_name")
conn_sl.execution_options(autocommit=True).execute(truncate_query)


q = select([table_1])

proxy = conn_p.execution_options(stream_results=True).execute(q)


while 'batch not empty':      
    batch = proxy.fetchmany(10000)  
    
    list1 = []

    if not batch:
        break
    
    for row in batch:
        d = dict(row.items())
        list1.append(d)    
    
    insert_stmt = table_2.insert().values(list1) 
    conn_sl.execute(insert_stmt)    


proxy.close()
```

**Comment by zzzeek** [comment, marked as answer]
hey -

if you did have COPY, you'd use the psycopg2 functions https://www.psycopg.org/docs/usage.html#copy .  Those are the fastest way to move data between two databases.

But without COPY, the code you have there is as good as you will get with SQLAlchemy(dont use the ORM for simple things that have to scale to millions of rows).   for the speed it's very likely the overhead is network / memory, since this is just fetching rows raw and there shouldn't be too much overhead for that.

the only way to make it faster would be to use raw psycopg2, if there's overhead in the compiler rendering out the INSERT..VALUES part of it ( at the same time, newer drivers like asyncpg or psycopg3 may be faster).       However, this might not be what's taking the time.  I would experiment with changing the size of the batch bigger / smaller.  see if not using a server side cursor helps or makes it worse.  if "table1" has many more columns that you don't actually need, try not using "select *", but we assume here it has the same columns.

you want to use the Python profiler to get a handle on where the time is being spent.  if it's spent in network overhead, then that's that, that's as fast as you can get the data over the wire.

if you have a simple standalone program, you can profile the whole thing at once at the commandline, see https://docs.python.org/3/library/profile.html#instant-user-s-manual .    we have a guide to profiling from a SQLAlchemy perspective at https://docs.sqlalchemy.org/en/14/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application

**Reply by zzzeek** [reply]
more thoughts, if the rows can be INSERTed in any order, you can parallelize the program using Python multiprocessing.   Or if it is indeed very IO bound,  doing the same but using asyncpg with asyncio tasks might get you the concurrency you need without using separate processes.

**Reply by ironsoul** [reply]
@zzzeek Thank you for the information, I will be testing on Monday. I really appreciate it.

**Comment by ironsoul** [comment]
After about a week of work, testing and optimizing, I am left with the following. 

I was working towards some ETL work but mostly lifting and shifting data from one postgresql DB to another one. I needed it to be dynamic in the way I pass tables to it, so that I could move tables easier. Please let me know what you think. 

  

    schema_s = "source schema"
    schema_d = "destination schema"


    source_creds = ('db connection string')
    destination_creds = ('db connection string')

    #same as destination - I use psycopg2 because it shaved off a few seconds off the insert time
    conn_ssyc = psycopg2.connect('db connection string')
    cur = conn_ssyc.cursor()


    #list of tables to move
    tbl_to_load = [
        'table_name_1',
        'table_name_2',
        'table_name_3',
        'table_name_4'
        ]

    for current_table in tbl_to_load:
        start_time_per = time.time()
        
        #create the needed engines
        conn_s = create_engine(source_creds, echo=False, echo_pool=False)
        conn_d = create_engine(destination_creds, echo=False, echo_pool=False)
        
        #get the columns in each current_table 
        df = pd.read_sql(
                    "SELECT \
                       column_name\
                    FROM information_schema.columns\
                        WHERE table_schema = '{}'\
                        AND table_name   = '{}'".format(schema_d, current_table), conn_d)
        cur = conn_ssyc.cursor()
        print(current_table)
        
        
        #creates a list of "%s" corresponding to the number of columns in each table. 
        #Eg table with 5 columns ("%s","%s","%s","%s","%s")

        inst_col = "("
        for i in df['column_name']:
            inst_col = inst_col + "%s,"
        inst_col = inst_col[:-1]    #removes the last comma
        inst_col = inst_col + ")"

        #defining the tables
        meta_p = MetaData(schema = schema_s)
        table_1 = Table(current_table,
                        meta_p,
                        autoload_with=conn_s
                       )


        meta_sl = MetaData(schema = schema_d)
        table_2 = Table(current_table,
                        meta_sl,
                        autoload_with=conn_d
                        )
        
        #truncate current table
        truncate_query = sqlalchemy.text("TRUNCATE TABLE {}.{}".format(schema_d, current_table))
        conn_d.execution_options(autocommit=True).execute(truncate_query)

        start_time= time.time()
        
        #steam the data
        q = select([table_1])
        proxy = conn_s.execution_options(stream_results=True).execute(q)
        
        end_time = time.time()
        total_time = end_time - start_time
        print("Time: ", total_time)
        
        batch_number = 0
        while 'batch not empty':  
            batch = proxy.fetchmany(10000)  # fetching batches of 10,000 rows at a time
            
            start_time= time.time()
            #batch_number is just to monitor progress
            batch_number = batch_number + 1
            print(batch_number)
            
            list1 = []

            if not batch:
                break

            for row in batch:
                d = dict(row.items())
                list1.append(d)    

            df = pd.DataFrame(list1)    

            tpls = [tuple(x) for x in df.to_numpy()]
            
            #Here is my insert, so far based on the table size (52 columns wide, 4.3m records, and full data size in csv is 3.5gb) I can move 10k records in 3-5 seconds per batch
            
            args_str = ','.join(cur.mogrify(inst_col, x).decode('utf-8') for x in tpls)
            cur.execute("INSERT INTO {}.{} VALUES ".format(schema_d, current_table) + args_str)
            conn_ssyc.commit()
            #4-5 sec or so for 10k
            
            end_time = time.time()
            total_time = end_time - start_time
            print("Time: ", total_time)

        cur.close()
        proxy.close()
        
        end_time = time.time()
        total_time = end_time - start_time_per
        print("Time: ", total_time)

**Reply by zzzeek** [reply]
it's not pretty but if it works, it works.

One thing I would change right off is that you can get the psycopg2 connection from the engine, no need to do psycopg2 connect, see https://docs.sqlalchemy.org/en/14/core/connections.html#working-with-the-dbapi-cursor-directly .  you can get the cursor from that and do the mogrify thing.

another thing I'd note however is that SQLAlchemy's psycopg2 dialect will use this mogrify system internally if you run an insert() with a list of dictionaries, using psycopg2's fast execution helpers.   it should be virtually identical performance-wise and would simplify the program a lot.   that said, you just have to get a job done and if it works it works.

**Reply by ironsoul** [reply]
Once again thanks for the advise, it has been really helpful. I am working on making it better and better.

I have a different question. I am looking at dropping a recreate tables. The reason I need to do this is because the source table columns change. The flow would be drop destination table and recreate base on source table. 

The source schema name is different to the destination. Is there away to use table_1 DDL to create the table_2 by change the schema its attached too?

I am looking at doing this to a list of tables like 'tbl_to_load' above. Can I solve this while using autoload_with.

```
meta_p = MetaData(schema = source_schema)
table_1 = Table(current_table,
                  meta_p,
                  autoload_with=conn_s
                 )
  
  
meta_sl = MetaData(schema = destination_schema)
table_2 = Table(current_table,
                  meta_sl,
                  autoload_with=conn_d
                  )
```

**Reply by CaselIT** [reply]
I think you can use the `schema_translate_map` feature to do that. https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=schema%20translate%20map#translation-of-schema-names

**Reply by ironsoul** [reply]
Thank you so much. This is what I was looking for. Will update here once I get it working.

**Reply by ironsoul** [reply]
After some testing, I haven't gotten this right. Can I pass this table create onto another connection. The source and destination are both in different databases. I would like to create the destination table base on what I find in the source database. The code snippet so far is:

```
    conn_source = create_engine(source_creds, echo=False, echo_pool=False)
    conn_destination = create_engine(destination_creds, echo=False, echo_pool=False)

conn_1 = conn_source.execution_options(stream_results=True, schema_translate_map={'public': "destination_schema"})

table_1.create(bind=conn_1, checkfirst=True)
```

but this all works on the source connection. How do I pass with onto the destination connection, so that I can create the tables?

**Reply by ironsoul** [reply]
Just ran the following and its working =D

```
conn2 = conn_destination.execution_options(schema_translate_map={'public': "destination_schema"})
table_1.create(bind=conn2, checkfirst=True)
```

**Reply by ironsoul** [reply]
@zzzeek @CaselIT  Thank you for all your help and input on this.
